Scripts/Runtime/Collections/UnsafeBuffer.cs

```diff
@@ -0,0 +1,550 @@
+            }
+            
+            /// <summary>
+            /// Creates a <see cref="LaneReader"/> instance locked to a specific lane index.
+            /// </summary>
+            /// <param name="laneIndex">The lane index to use</param>
+            /// <returns>A <see cref="LaneReader"/> instance</returns>
+            public LaneReader CreateReaderForLane(int laneIndex)
+            {
+                Assert.IsTrue(laneIndex < m_BufferInfo->LaneCount && laneIndex >= 0);
+                
+                Lane* lane = m_Lanes + laneIndex;
+                return new LaneReader(lane, m_BufferInfo);
+            }
+        }
+        
+        /// <summary>
+        /// A lightweight reader instance locked to a specific lane.
+        /// <see cref="LaneReader"/>'s manage their own state so that multiple readers can read from a given lane at
+        /// the same time.
+        /// </summary>
+        [BurstCompatible]
+        public struct LaneReader
+        {
+            [NativeDisableUnsafePtrRestriction] private readonly BufferInfo* m_BufferInfo;
+            [NativeDisableUnsafePtrRestriction] private readonly Lane* m_Lane;
+            
+            private BlockInfo* m_CurrentReadBlock;
+            private byte* m_EndOfCurrentBlock;
+            private byte* m_ReaderHead;
+            
+            /// <summary>
+            /// How many elements are available to read on this lane
+            /// </summary>
+            public int Count
+            {
+                get => m_Lane->Count;
+            }
+            
+            internal LaneReader(Lane* lane, BufferInfo* bufferInfo)
+            {
+                m_Lane = lane;
+                m_BufferInfo = bufferInfo;
+                
+                m_CurrentReadBlock = m_Lane->FirstBlock;
+                m_ReaderHead = (m_CurrentReadBlock != null) 
+                    ? m_CurrentReadBlock->Data
+                    : null;
+                m_EndOfCurrentBlock = m_ReaderHead + m_BufferInfo->BlockSize;
+            }
+
+            private void CheckForEndOfBlock()
+            {
+                if (m_ReaderHead < m_EndOfCurrentBlock)
+                {
+                    return;
+                }
+                
+                m_CurrentReadBlock = m_CurrentReadBlock->Next;
+                m_ReaderHead = (m_CurrentReadBlock != null) 
+                    ? m_CurrentReadBlock->Data
+                    : null;
+                m_EndOfCurrentBlock = m_ReaderHead + m_BufferInfo->BlockSize;
+            }
+            
+            /// <summary>
+            /// Reads the next element from the reader
+            /// </summary>
+            /// <returns>The next element</returns>
+            public ref T Read()
+            {
+                //Otherwise nothing has been written
+                //TODO: Write test for this
+                Assert.IsTrue(Count > 0 && m_CurrentReadBlock != null && m_ReaderHead != null);
```

Comment:
Do we always want these asserts or should they be toggled via `ENABLE_UNITY_COLLECTIONS_CHECKS`

Reply:
Sure I'll add the checks.